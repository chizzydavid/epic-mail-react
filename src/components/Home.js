import React, { Component } from 'react'
import { Link } from 'react-router-dom';

const Landing = () => {
  return (
    <React.Fragment>
      <div >
        <h1 className="sub-text">HOMEPAGE</h1>
      </div>
   </React.Fragment>
  )
}

export default Landing
