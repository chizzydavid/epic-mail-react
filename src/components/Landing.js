import React, { Component } from 'react'
import { Link } from 'react-router-dom';

const Landing = () => {
  return (
    <React.Fragment>
      <div id="background" className="header-text">
        <p className="banner-header">Welcome to <strong>Epic Mail</strong></p>
        <p className="sub-text">An exceptional email delivery service</p>

        <div id="buttons">
          <Link className="start-btn" to="/sign-up">SIGN UP</Link>
          <a className="start-btn transparent" href="#about">LEARN MORE</a>        
          <p>Already have an account? <Link className="sign-link" to="/login">Log In</Link></p>  
        </div>
      </div>

      <section className="" id="about"> 
        <div className="about-text"> 
          <h2 className="subhead">Getting Started</h2>
            <p> Epic Mail is an excellent application designed to make sending and receiving email messages a lot easier and fun. Built using the latest web technologies and current best practices to optimize performance and ensure a great user experience.
          </p>
        </div>

        <div className=" about-row">
          <div className="about-item" >
            <div className=""><i className="fa fa-sign-in"></i></div>
            <div>
              <h3>Create Account</h3>
              <p>Click the sign up button above to create an account, fill the form with the required information, you could also choose to upload a picture of yourself or not. And then click submit. </p>
            </div> 
          </div>

          <div className="about-item" >
            <div className=""><i className="fa fa-user"></i></div>
            <div>
              <h3>Login</h3>
              <p>Login in with the email address and password used in creating your account. Then go on to enjoy all the incredible features of this application. </p>
            </div> 
          </div>

          <div className="about-item" >
            <div className=""><i className="fa fa-envelope"></i></div>
            <div>
              <h3>Send Messages</h3>
              <p>Create and send emails to either an individual or a select group of persons. If you wish you can save said message as a draft and come back later on to finish. </p>
            </div> 
          </div>


        </div>
      </section>
    </React.Fragment>
  )
}

export default Landing
