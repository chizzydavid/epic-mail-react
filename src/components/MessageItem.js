import React from 'react'
import PropTypes from 'prop-types';

const MessageItem = (props) => {
  const imgUrl = `https://res.cloudinary.com/chizzydavid/image/upload/`;
  
  const {
    message_id, subject, status, email, sender_id, message, photo 
  } = props

	const id = Number(message_id);
  const msgBody = message.length >= 70 ? `${message.substr(0, 70)}...` : message;
  const type = Number(2) === sender_id ? 'sent' : 'received';
  const msgStatus = status.replace(status[0], status[0].toUpperCase());
  const userImg = `${imgUrl}${photo}`;
  return (
		<div className="message" data-message-id={id} data-message-type={type} data-message-status={status}>

			<div className="message-img">
				<img src={userImg} id="msg-sender-img"/>
			</div>

			<div className="message-text">
				<h4 className={status == 'unread' ? 'message-title unread' : 'message-title'}> {subject}	</h4>
	
				<div className="msg-body"> 
					<p className="msg-excerpt">{msgBody}</p>
					<div className="msg-details">
						<p>Status: <span className={'status-' + status + 'msg-status'}>{msgStatus} </span> </p>

						<div className="msg-buttons">
							{status === 'draft' ? (
								<React.Fragment>
									<i id={id} data-draft-receiver={email} className="edit-draft fa fa-edit"></i>
									<p className="hide">  {message}  </p>									
								</React.Fragment>
							) : '' }
							
							{type === 'sent' && status !== 'draft'
							? (<i id={id} className="retract-message fa fa-undo"></i>) : '' }
							<i id="${id}" data-message-type="${type}" className="delete-message fa fa-trash"></i>
						</div>            
					</div> 
				</div>    
			</div>
		</div>
  )
};

MessageItem.propTypes = {
  message_id: PropTypes.number.isRequired,
  subject: PropTypes.string.isRequired,
  email: PropTypes.string,
  sender_id: PropTypes.number.isRequired,
  message: PropTypes.string.isRequired,
  status: PropTypes.string,
  photo: PropTypes.string.isRequired
}

export default MessageItem
