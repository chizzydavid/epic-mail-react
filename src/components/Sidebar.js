import React, { Component } from 'react'
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { setActiveMsgTab } from '../store/actions/messageActions';

export class Sidebar extends Component {
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick(e) {
    if (e.target.tagName === 'I') {
      this.props.handleClick(e.target.parentElement.id);
      return;
    }
    this.props.handleClick(e.target.id);
  }

	render() {
    const { activeMsgTab } = this.props;

		return (
      <aside className="messages-nav">
        <Link to="/compose">
          <p id="compose-btn">Compose <i className="fa fa-plus"> </i></p>
        </Link>

        <div id="nav-wrapper">
          <div id="btn-wrapper"> <i id="inbox-nav-btn" className="fa fa-caret-down"> </i></div>
          <ul>
            <li onClick={this.handleClick} id="all" className={activeMsgTab === 'all' ? 'active' : ''}>
              <i className="fa fa-folder"> </i>All Received</li>

            <li onClick={this.handleClick} id="unread" className={activeMsgTab === 'unread' ? 'active' : ''}>
              <i className="fa fa-envelope"> </i>Unread</li>

            <li onClick={this.handleClick} id="read" className={activeMsgTab === 'read' ? 'active' : ''}>
              <i className="fa fa-envelope-open"> </i>Read</li>

            <li onClick={this.handleClick} id="sent" className={activeMsgTab === 'sent' ? 'active' : ''}>
              <i className="fa fa-send"> </i>Sent</li>

            <li onClick={this.handleClick} id="draft" className={activeMsgTab === 'draft' ? 'active' : ''}>
              <i className="fa fa-edit"> </i>Drafts</li>
          </ul>
        </div>
      </aside>
		)
	}
}

Sidebar.propTypes = {
  setActiveMsgTab: PropTypes.func.isRequired,
  activeMsgTab: PropTypes.string.isRequired,
  handleClick: PropTypes.func.isRequired
}

const mapStateToProps = (state) => ({
  activeMsgTab: state.messages.activeMsgTab,
});

const mapDispatchToProps = {
  setActiveMsgTab
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Sidebar);

