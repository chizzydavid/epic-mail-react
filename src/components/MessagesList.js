import React from 'react'
import PropTypes from 'prop-types';
import MessageItem from './MessageItem';

const MessageList = (props) => {
  const { 
    userId, isLoading, messages, feedback, errors, handleDelete
  } = props;

  const formFeedback = feedback ? feedback : errors.error ? errors.error : '';

  const allMessages = messages.map(({message_id, subject, sender_id, message, photo, status}) => 
    <MessageItem 
      userId={userId}
      key={message_id}
      message_id={message_id}
      sender_id={sender_id}
      subject={subject}
      message={message}
      photo={photo}
      status={status}
      handleDelete={handleDelete}
    />
  )

  return (
    <main className="message-wrapper">
      <div className={isLoading ? '' : 'hide'} id="loader">
        <div className="sk-wave">
          <div className="sk-rect sk-rect1"></div>
          <div className="sk-rect sk-rect2"></div>
          <div className="sk-rect sk-rect3"></div>
          <div className="sk-rect sk-rect4"></div>
          <div className="sk-rect sk-rect5"></div>
        </div>
      </div>
      <div>
        <p className={formFeedback ? '' : isLoading ? 'hide' : ''} id="feedback">
          {formFeedback}
        </p>
      </div>

      <div className="message-container"> {allMessages} </div>
    </main>
  )
}

MessageList.propTypes = {
  userId: PropTypes.number.isRequired,
  messages: PropTypes.array.isRequired,
  isLoading: PropTypes.bool.isRequired,
  feedback: PropTypes.string.isRequired,
  errors: PropTypes.object.isRequired,
  handleDelete: PropTypes.func.isRequired
}

export default MessageList;
