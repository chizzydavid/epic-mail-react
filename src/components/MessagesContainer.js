import React, { Component } from 'react'
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';
import Sidebar from './Sidebar';
import MessagesList from './MessagesList';
import { getMessages } from '../store/actions/messageActions';

export class MessagesContainer extends Component {
  componentDidMount() {
    this.props.getMessages('all');
  }

  handleTabChange = (category) => {
    this.props.getMessages(category);
  }

  render() {
    const { 
      isLoading, activeMsgTab, feedback, allReceived, unread, read, sent, drafts, errors 
    } = this.props.messages;
    const { isAuthenticated } = this.props;
    if (!isAuthenticated) {
      return <Redirect to="/login" />;
    }  
    let headerName;
    let sortedMessages;

    switch (activeMsgTab) {
      case 'all':
        headerName = 'All Received Messages';
        sortedMessages = allReceived;
        break;
      case 'read':
        headerName = 'Unread Messages';
        sortedMessages = unread;
        break;      
      case 'unread':
        headerName = 'Read Messages';
        sortedMessages = read;
        break;  
      case 'sent':
        headerName = 'Sent Messages';
        sortedMessages = sent;
        break;  
      case 'draft':
        headerName = 'Drafts';
        sortedMessages = drafts;
        break;
      default:
        break;
    }

		return (
			<div id="messages-wrapper">
        <h2 className="msg-header">{headerName}</h2>

        <div className="main-container">
          <Sidebar 
            handleClick={this.handleTabChange}
          />
          <MessagesList
            isLoading={isLoading}
            feedback={feedback}
            errors={errors}
            messages={sortedMessages}
          />
        </div>
			</div>
		)
	}
}

MessagesContainer.propTypes = {
  messages: PropTypes.object.isRequired,
  getMessages: PropTypes.func.isRequired,
  isAuthenticated: PropTypes.bool.isRequired
}

const mapStateToProps = (state) => ({
  messages: state.messages,
  isAuthenticated: state.auth.isAuthenticated
});

const mapDispatchToProps = {
  getMessages
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MessagesContainer);

