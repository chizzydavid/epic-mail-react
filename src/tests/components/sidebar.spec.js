import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import { Sidebar } from '../../components/Sidebar';

function shallowSetup() {
  const props = {
    activeMsgTab: 'read',
    setActiveMsgTab: () => {},
    handleClick: () => {},
  }
  const enzymeWrapper = shallow(<Sidebar {...props} />);

  return {
    props,
    enzymeWrapper
  };
}

describe('Sidebar component', () => {
  beforeEach(() => {
    sinon.spy(Sidebar.prototype, 'handleClick');
  });

  afterEach(() => {
    Sidebar.prototype.handleClick.restore();
  });

  it('should render sidebar component', () => {
    const { enzymeWrapper } = shallowSetup();
    expect(enzymeWrapper).toMatchSnapshot();
  });

  it('should call handleclick ', () => {
    const { enzymeWrapper } = shallowSetup();

    enzymeWrapper.find('li').first().simulate('click', {
      target: {
        id: 'read',
      },
    });
    expect(Sidebar.prototype.handleClick.calledOnce).toBe(true);
  });

  it('should call handleclick ', () => {
    const { enzymeWrapper } = shallowSetup();

    enzymeWrapper.find('li').first().simulate('click', {
      target: {
        tagName: 'I',
        parentElement: { id: 'read' }

      },
    });
    expect(Sidebar.prototype.handleClick.calledOnce).toBe(true);
  });

});

