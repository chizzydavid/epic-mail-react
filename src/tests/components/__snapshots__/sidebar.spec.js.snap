// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar component should render sidebar component 1`] = `ShallowWrapper {}`;
